### ISSUE TITLE
SSO not working using Soicaluser module

### ISSUE THREAD

**Issue by abdulwaheed18** [Digipalvelutehdas/MPASSid-proxy#72]
## Use Case

Downloaded and Integrated social login module with my Shibboleth Idp and able to do authentication via Facebook but I am facing an issue when I try to access another application which is protected by another SP federated with same Idp. When I try to access it, It is redirecting me to the Facebook landing page for authentication even when I have active Shibboleth Idp Session.

Note: At first glance, it looks like SSO is working but after checking network log I realized that second time authentication is happening because of existing facebook session and it is not using JSESSION ID.

## Reproduce Steps:

Assume we have two different hosts (host1.domain.com & host2.domain.com) and two applications app1 and app2 which is protected by different Sp and a single Idp which is internally using socialuser plugin for Facebook.

host1.domain.com and host2.domain.com both are federated with same Shibboleth Idp.

1. Hit host1.domain.com/app1 and it will redirect to the facebook page for authentication, after authentication I can access my app1.
2. Open new tab and logout from facebook.
3. Hit host2.domain.com/app2 and it is again redirecting me back to the facebook login page even though I have active JSESSIONID and idp_shib_session.

Ideally, I should not ask for authentication at the second time which is not happening right now.

**Comment by hjmikkon** [comment]
Thanks for the report!

Indeed, the SSO feature was disabled for all the other SocialUser login flows, except SocialUserOpenIDConnect. The feature is now enabled by default in the master branch, and one can configure it in the following way:

flows/authn/SocialUser/socialuser-authn-beans.xml:

...
	<bean id="ValidateExternalAuthentication"
		class="net.shibboleth.idp.authn.impl.ValidateExternalAuthentication"
		scope="prototype" p:classifiedMessages-ref="shibboleth.authn.SocialUser.ClassifiedMessageMap"
		p:resultCachingPredicate="**<VALUE HERE>**" />
...

Replace **<VALUE HERE>** with _false_ to disable SSO, or _true_ to enable it.

Does this work ok for you?

**Comment by abdulwaheed18** [comment]
Thanks for the reply.

I was not sure about this setting so I've enabled it directly by changing the code itself. It was blocked by 

>      authenticationContext.setResultCacheable(false);
in ExtractSocialPrincipalsFromSubject file.

Now, I am facing another issue with social login flow i.e SSO is happening but Idp is not able to send attribute statement at the second time.

Basically, I have added attribute-resolver-social which release all the attribute statement like providerId, sn, mail and so on by fetching it from the SocialUserPrincipal but at the time of another request (SSO), Idp checks the previous session already exist and it post the SAML IdP response to SP without any attribute Statement. Seems like socialContext is not present at the second time.

**Comment by abdulwaheed18** [comment]
I have one more question, Ignore it if you don't find it relevant but I am curious about it.

You've added social login flow to support login from the various network like Google, facebook, twitter and so on but is it possible to enable all the flow at the same time?

I have seen your discovery module but the configuration and decision will be done at Idp Side.

I just need to know is it possible to enable all flows at the same time and can trigger particular flow by the SP?

Basically, Idp.authn.flow will be having

> Idp.authn.flow = Password|SocialuserFacebook|SocialUserGoogle

and which one to trigger can decide ar the time of hitting SP?

**Comment by hjmikkon** [comment]
Your observation "Seems like socialContext is not present at the second time." is correct: at the moment you cannot access any other information except userId in the attribute resolver if you have enabled SSO.

Background: in the MPASSid instance itself, we have disabled SSO in the Shibboleth IdP, because we always forward the end-user to be authenticated at SocialUser or other authenticator-flows. They provide SSO, and we rely on that. In the end we also always have the contexts available during the attribute resolution phase.

One way to access the SocialUserPrincipals would be via Subject (authnContext.getAuthenticationResult().getSubject()), but in order to make that working we would need to implement a principal serializer (see for instance net.shibboleth.idp.authn.principal.impl.UsernamePrincipalSerializer). That implementation doesn't exist yet, but we can work on it if you see it important?

Regarding the other question, you can for instance configure custom SAML authentication context class references to your flows in the /opt/shibboleth-idp/conf/authn/general-authn.xml file. Something like this:

```
        <bean id="authn/SocialUserFacebook" parent="shibboleth.AuthenticationFlow"
            p:nonBrowserSupported="false"
            p:passiveAuthenticationSupported="true"
            p:forcedAuthenticationSupported="true">
            <property name="supportedPrincipals">
                <util:list>
                    <bean parent="shibboleth.SAML2AuthnContextClassRef"
                        c:classRef="urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport" />
                    <bean parent="shibboleth.SAML2AuthnContextClassRef"
                        c:classRef="urn:testing:facebook" />
                </util:list>
            </property>
        </bean>

```

Whenever an SP would require _urn:testing:facebook_ context class reference in the authnRequest, the IDP should automatically select facebook if that's the only flow supporting the requested context class.

If you're using Shibboleth SP, you can request specific context class via authnContextClassRef -parameter in the Session Initiator (../Shibboleth.sso/Login).

**Comment by abdulwaheed18** [comment]
Thanks once again for the clarification.

IMHO, If we are enabling SSO then I guess, Ideally it should returns all the attributes statement in each SAML IdP response. I'll look into the "net.shibboleth.idp.authn.principal.impl.UsernamePrincipalSerializer" to get the better Idea.

Regarding UserId, How can I retrieve it in attribute-resolver-social.xml file? I tried to fetch it but didn't get any luck.

> <resolver:AttributeDefinition xsi:type="ad:Script" id="userId">
>         <resolver:AttributeEncoder xsi:type="enc:SAML1String" name="urn:mace:dir:attribute-def:userId" encodeType="false" />
>         <resolver:AttributeEncoder xsi:type="enc:SAML2String" name="urn:oid:2.16.840.1.113730.3.1.223" friendlyName="userId" encodeType="false" />
>         <ad:Script><![CDATA[
>           authnContext = resolutionContext.getParent().getSubcontext("net.shibboleth.idp.authn.context.AuthenticationContext");
>          // MISSING_CODE_TO_FETCH ID
>         ]]></ad:Script>
>     </resolver:AttributeDefinition>
> 

and one more thing,  there is one minor bug in attribute-resolver-social.xml file
When we enable multiple flows in IDP and choose flow as Password for authentication then you will see Exception in idp_warn.log file saying "null does not have function getEmail." and it is happening because when we choose Password flow then social context will always be null and  **mail.addValue(socialContext.getEmail());** will throw null pointer Exception. I have fixed it by adding if statement.

>  <resolver:AttributeDefinition xsi:type="ad:Script" id="displayName">
>         <resolver:AttributeEncoder xsi:type="enc:SAML1String" name="urn:mace:dir:attribute-def:displayName" encodeType="false" />
>         <resolver:AttributeEncoder xsi:type="enc:SAML2String" name="urn:oid:2.16.840.1.113730.3.1.241" friendlyName="displayName" encodeType="false" />
>         <ad:Script><![CDATA[
>           authnContext = resolutionContext.getParent().getSubcontext("net.shibboleth.idp.authn.context.AuthenticationContext");
>           socialContext = authnContext.getSubcontext("fi.okm.mpass.idp.authn.context.SocialUserContext");
> 		   if(null != socialContext){
> 			displayName.addValue(socialContext.getDisplayName());
> 			}
>         ]]></ad:Script>
>     </resolver:AttributeDefinition>

and the last point regarding triggering specific flow from SP side: I tried the same using sessionInitiator but there is one issue with that approach i.e. If you have enabled multiple ApplicationOverrides in your shibboleth2.xml then sessionInitiator will add the same authContextRef to each entityId.

My idea is to have discoveryService at SP side having drop down items having google|facebook|password|Salesforce and so on and if a user clicks on facebook then it should trigger facebook flow which is not happening.

As of now, I am triggering it by adding authnContextref in apache's ssl.conf file and the decision is based on URL.

ShibRequestSetting applicationId dd555ba2-319b-435d-9ee7-3d6e94095101
ShibRequestSetting authnContextClassRef urn:oasis:names:tc:SAML:2.0:ac:classes:TimeSyncToken

TimeSyncToken is matching to facebook profile in general_authn.xml and applicationId is matching to specific ApplicationOverrides in shibboleth2. I am not sure if this is the right approach but it is working for me.

I guess ideal way would be from SP side which is not happening and I am still exploring it.

**Comment by jalauros** [comment]
> I was not sure about this setting so I've enabled it directly by changing the code itself. It was blocked by
> 
>      authenticationContext.setResultCacheable(false);
> 
> in ExtractSocialPrincipalsFromSubject file.

I placed that call there to ensure the attributes will be resolvable for our use case. As you have already noted, the context is lost for the second time around. Of course there are better ways to handle that than hardcoding it but that is the history behind that.

**Comment by abdulwaheed18** [comment]
there are better ways to handle that hardcoding it but that is the history behind that.

Yes, I can understand that Implementing social login is not that straight forward. there are lots of challenge to face while implementing it and I really appreciate all the hard work that you guys have done to make it available. :+1: 

After enabling SSO, Now we face multiple challenges as already mentioned above:

1. **SocialUserPrinicpal gets lost for the second time:** Is there any quick solution using which we can retrieve the all the social user details at the second time too? are you guys planning to support this feature?

2. **UserId retrieval**: As of now, I can see that we are storing userId in the subject. I believe, there is some way using which we can retrieve it in attribute-resolver-social.xml file.

>   subject.getPrincipals().add(new UsernamePrincipal(userId));

I tried various approach to get the userId but no luck so far. Any help would be appreciated :)

**Comment by hjmikkon** [comment]
> SocialUserPrinicpal gets lost for the second time: Is there any quick solution using which we can retrieve the all the social user details at the second time too? are you guys planning to support this feature?

I just pushed the SocialUserPrincipalSerializer to the master branch, which provides a solution to your issue.

First, you need to enable the serializer. I did it myself by adding the following to _/opt/shibboleth-idp/conf/global.xml_:

```
    <bean id="shibboleth.PrincipalSerializers"
            class="org.springframework.beans.factory.config.ListFactoryBean">
        <property name="sourceList">
            <list>
                <bean class="fi.okm.mpass.idp.authn.principal.impl.SocialUserPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.UsernamePrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.LDAPPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.duo.impl.DuoPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.IdPAttributePrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.PasswordPrincipalSerializer"
                    p:dataSealer="#{'%{idp.sealer.storeResource:}'.trim().length() > 0 ? getObject('shibboleth.DataSealer') : null}" />
            </list>
        </property>
    </bean>
```

The other serializers are the same ones that are defined in the _/opt/shibboleth-idp/system/conf/general-authn-system.xml_ (see bean **shibboleth.DefaultPrincipalSerializers**).

After that, you're able to read the SocialUserPrincipals' contents in the attribute resolver. Also in the SSO sessions. See the example below for reading the information for _providerId_ attribute. The same logic should work for other attributes too, like email, firstName, lastName, if they are provided by the social media provider.

```
        <ad:Script><![CDATA[
          authnContext = resolutionContext.getParent().getSubcontext("net.shibboleth.idp.authn.context.AuthenticationContext");
          subject = authnContext.getAuthenticationResult().getSubject();
          principals = subject.getPrincipals(Java.type("fi.okm.mpass.idp.authn.principal.SocialUserPrincipal").class);
          iterator = principals.iterator();
          while (iterator.hasNext()) {
              principal = iterator.next();
              if ("providerId".equals(principal.getType())) {
                  providerId.addValue(principal.getValue());
              }
          }
	]]></ad:Script>
```

> UserId retrieval: As of now, I can see that we are storing userId in the subject. I believe, there is some way using which we can retrieve it in attribute-resolver-social.xml file.

As the userId is set to UsernamePrincipal (as you have noticed), you can set it in the following way:

```
    <AttributeDefinition id="uid" xsi:type="PrincipalName">
        <AttributeEncoder xsi:type="SAML1String" name="urn:mace:dir:attribute-def:uid" encodeType="false" />
        <AttributeEncoder xsi:type="SAML2String" name="urn:oid:0.9.2342.19200300.100.1.1" friendlyName="uid" encodeType="false" />
    </AttributeDefinition>
```

Hopefully you can get the social + SSO combination working now! Let us know if you have any questions or comments?

**Comment by abdulwaheed18** [comment]
That was really quick. Thanks!

I did all the changes as you have suggested including global.xml

```
<bean id="shibboleth.PrincipalSerializers"
            class="org.springframework.beans.factory.config.ListFactoryBean">
        <property name="sourceList">
            <list>
                <bean class="fi.okm.mpass.idp.authn.principal.impl.SocialUserPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.UsernamePrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.LDAPPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.IdPAttributePrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.PasswordPrincipalSerializer"
                    p:dataSealer="#{'%{idp.sealer.storeResource:}'.trim().length() > 0 ? getObject('shibboleth.DataSealer') : null}" />
            </list>
        </property>
    </bean>
```

 but with that SSO does not work me. I mean, I was still losing the social context at second request but When I added

> bean class="fi.okm.mpass.idp.authn.principal.impl.SocialUserPrincipalSerializer"

this into **general-authn-system.xml** then SSO started working absolutely fine and I am getting social context and I guess, Adding anything into the system is not a good practice as it will get deleted after upgrade.

**Comment by hjmikkon** [comment]
It's weird that you had to modify any file in the _system/_ directory. In the **general-authn-system.xml** there is an existing bean:

```
    <!-- The serializer machinery injected into authentication flows by default. -->
    <bean id="shibboleth.DefaultAuthenticationResultSerializer"
            class="net.shibboleth.idp.authn.impl.DefaultAuthenticationResultSerializer">
        <property name="principalSerializers">
            <ref bean="#{getObject('shibboleth.PrincipalSerializers') != null
                ? 'shibboleth.PrincipalSerializers' : 'shibboleth.DefaultPrincipalSerializers'}" />
        </property>
        <property name="genericPrincipalSerializer.symbolics">
            <ref bean="#{getObject('shibboleth.PrincipalSymbolics') != null
                ? 'shibboleth.PrincipalSymbolics' : 'shibboleth.DefaultPrincipalSymbolics'}" />
        </property>
    </bean>
```

... which uses the bean called **shibboleth.PrincipalSerializers** if it exists, and **shibboleth.DefaultPrincipalSerializers** if it doesn't. By specifying the **shibboleth.PrincipalSerializers** in _/opt/shibboleth-idp/conf/global.xml_ should then mean that it exist also here, and that's the case in my installation. I didn't change anything in the _system/_ -directory.

I would double-check the file and bean names for possible typos?

**Comment by abdulwaheed18** [comment]
Here is the DefaultAuthenticationResultSerializer bean in general-authn-system.xml file:

```
 <bean id="shibboleth.DefaultAuthenticationResultSerializer"
            class="net.shibboleth.idp.authn.impl.DefaultAuthenticationResultSerializer">
        <property name="principalSerializers">
            <list>
                <bean class="net.shibboleth.idp.authn.principal.impl.UsernamePrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.LDAPPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.IdPAttributePrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.PasswordPrincipalSerializer"
                    p:dataSealer-ref="shibboleth.DataSealer" />
            </list>
        </property>
        <property name="genericPrincipalSerializer.symbolics">
            <map>
                <!-- Custom principal types. -->
                <entry key="net.shibboleth.idp.saml.authn.principal.AuthnContextClassRefPrincipal" value="1" />
                <entry key="net.shibboleth.idp.saml.authn.principal.AuthnContextDeclRefPrincipal" value="2" />
                <entry key="net.shibboleth.idp.saml.authn.principal.AuthenticationMethodPrincipal" value="3" />
                <entry key="javax.security.auth.kerberos.KerberosPrincipal" value="4" />
                
                <!-- SAML 2 authn context classes. -->
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:InternetProtocol" value="100" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:InternetProtocolPassword" value="101" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:Kerberos" value="102" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:MobileOneFactorUnregistered" value="103" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:MobileTwoFactorUnregistered" value="104" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:MobileOneFactorContract" value="105" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:MobileTwoFactorContract" value="106" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:Password" value="107" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport" value="108" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:PreviousSession" value="109" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:X509" value="110" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:PGP" value="111" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:SPKI" value="112" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:XMLDSig" value="113" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:Smartcard" value="114" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:SmartcardPKI" value="115" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:SoftwarePKI" value="116" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:Telephony" value="117" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:NomadTelephony" value="118" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:PersonalTelephony" value="119" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:AuthenticatedTelephony" value="120" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:SecureRemotePassword" value="121" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:TLSClient" value="122" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:TimeSyncToken" value="123" />
                <entry key="urn:oasis:names:tc:SAML:2.0:ac:classes:unspecified" value="124" />

                <!-- SAML 1 authn methods -->
                <entry key="urn:ietf:rfc:1510" value="201" />
                <entry key="urn:oasis:names:tc:SAML:1.0:am:HardwareToken" value="202" />
                <entry key="urn:oasis:names:tc:SAML:1.0:am:X509-PKI" value="203" />
                <entry key="urn:oasis:names:tc:SAML:1.0:am:PGP" value="204" />
                <entry key="urn:oasis:names:tc:SAML:1.0:am:SPKI" value="205" />
                <entry key="urn:oasis:names:tc:SAML:1.0:am:XKMS" value="206" />
                <entry key="urn:ietf:rfc:3075" value="207" />
                <entry key="urn:ietf:rfc:2945" value="208" />
                <entry key="urn:ietf:rfc:2246" value="209" />
                <entry key="urn:oasis:names:tc:SAML:1.0:am:unspecified" value="210" />
                
                <!-- Community-defined assurance profiles. -->
                <entry key="http://id.incommon.org/assurance/silver" value="300" />
                <entry key="http://id.incommon.org/assurance/bronze" value="301" />
            </map>
        </property>
    </bean>
```

and this is what I've tried to add in global.xml file:

	   ```<bean id="shibboleth.PrincipalSerializers"
            class="org.springframework.beans.factory.config.ListFactoryBean">
        <property name="sourceList">
            <list>
                <bean class="fi.okm.mpass.idp.authn.principal.impl.SocialUserPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.UsernamePrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.LDAPPrincipalSerializer" />
                <bean class="net.shibboleth.idp.authn.principal.impl.IdPAttributePrincipalSerializer" />
                 <bean class="net.shibboleth.idp.authn.principal.impl.PasswordPrincipalSerializer"
                    p:dataSealer-ref="shibboleth.DataSealer" />
            </list>
        </property>
    </bean>
```

**Comment by hjmikkon** [comment]
For me it looks like you have quite old version of _general-authn-system.xml_ file in use?

This is the original dev-ticket for the required functionality: https://issues.shibboleth.net/jira/browse/IDP-740

It has been committed to the template xml-file in 26.2.2016, see:

https://git.shibboleth.net/view/?p=java-identity-provider.git;a=history;f=idp-conf/src/main/resources/system/conf/general-authn-system.xml;h=dc00594989d9e1c4a6f5356366e961372b3ca821;hb=HEAD

**Comment by abdulwaheed18** [comment]
Thanks once again for all the help!

Yes, you are right! I am using Idp 3.2.1 and I guess the changes are in the later version. I have noted it down and will do the changes in global once I upgrade my local Idp. 

I've one last query and it is more related to activating multiple Social login flows from shibboleth SP side.

Let's assume I have Discovery Service enabled on my Shibboleth Sp and have three different metadata:
  1. Shibboleth Idp (**EntityId: https://idp.com/idp/shibboleth**)
  2. Facebook (**EntityId: https://idp.com/idp/facebook**, Same as Shibboleth Idp metadata but changed entity into It)
  3. Google ( **EntityId: https://idp.com/idp/google**, by updating entity Id)

If I hit anyone of the Idp(IdP or Facebook or Google ) from drop down list it will go to the password flow of Shibboleth Idp as we are not passing any **AuthnContextClassRef**.

Now my question: Is it possible to add particular **AuthnContextClassRef** to speciifc entityId i.e can we pass authnContextClassRef **urn:social:facebook** to (**EntityId: https://idp.com/idp/shibboleth**)? I tried adding SessionInitiator

```
<SessionInitiator type="SAML2" isDefault="false" Location="/Login" entityID="https://idp.com/idp/facebook">
	<samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" Version="2.0" IssueInstant="2012-01-01T00:00:00Z">
	    <samlp:RequestedAuthnContext Comparison="exact" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">
	<saml:AuthnContextClassRef>urn:social:facebook</saml:AuthnContextClassRef>
	    </samlp:RequestedAuthnContext>
	</samlp:AuthnRequest>
</SessionInitiator>
```
When I add above sessionInitiator then It is adding **AuthnContextClassRef** to each entityId and hitting facebook or google will choose facebook flow in Shibbolet IdP.

So, Is it possible to trigger specific authn flow from SP side?

Thanks in Advance!

**Comment by mjpmachado** [comment]
Hello.
I'm trying to implement the Facebook authentication flow. 
I can login into a service without any error and get the social user details/ attributes delivered by the Identity Provider. I'm using the IdP version 3.3.2 .
If I try to access to other service that is protected by the same Identity Provider is expected to get all the social user details again.
I have the following error:

`ERROR [net.shibboleth.idp.profile.impl.ResolveAttributes:299] - Profile Action ResolveAttributes: Error resolving attributes
net.shibboleth.idp.attribute.resolver.ResolutionException: Attribute Definition 'mail': unable to execute script        at net.shibboleth.idp.attribute.resolver.ad.impl.ScriptedAttributeDefinition.doAttributeDefinitionResolve(ScriptedAttributeDefinition.java:190)
Caused by: javax.script.ScriptException: TypeError: null has no such function "getEmail" in <eval> at line number 13
        at jdk.nashorn.api.scripting.NashornScriptEngine.throwAsScriptException(NashornScriptEngine.java:470)
Caused by: jdk.nashorn.internal.runtime.ECMAException: TypeError: null has no such function "getEmail"
        at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)`

I implement the suggested ideais mentioned above but without success.

Steps that I have done:
1.  flows/authn/SocialUser/socialuser-authn-beans.xml:
`	<bean id="ValidateExternalAuthentication"
        	class="net.shibboleth.idp.authn.impl.ValidateExternalAuthentication"
        	scope="prototype" p:classifiedMessages-ref="shibboleth.authn.SocialUser.ClassifiedMessageMap"
        	p:resultCachingPredicate="true" />`

2.  /opt/shibboleth-idp/conf/authn/general-authn.xml
`<bean id="authn/SocialUserFacebook" parent="shibboleth.AuthenticationFlow"
            p:nonBrowserSupported="false"
            p:passiveAuthenticationSupported="true"
            p:forcedAuthenticationSupported="true">
            <property name="supportedPrincipals">
                <util:list>
                    <bean parent="shibboleth.SAML2AuthnContextClassRef"
                        c:classRef="urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport" />
                    <bean parent="shibboleth.SAML2AuthnContextClassRef"
                        c:classRef="urn:testing:facebook" />
                </util:list>
            </property>
        </bean>
`

3. /opt/shibboleth-idp/conf/global.xml
`	<bean id="shibboleth.PrincipalSerializers" class="org.springframework.beans.factory.config.ListFactoryBean">
        	<property name="sourceList">
           	<list>
                	<bean class="fi.okm.mpass.idp.authn.principal.impl.SocialUserPrincipalSerializer" />
                	<bean class="net.shibboleth.idp.authn.principal.impl.UsernamePrincipalSerializer" />
                	<bean class="net.shibboleth.idp.authn.principal.impl.LDAPPrincipalSerializer" />
                	<bean class="net.shibboleth.idp.authn.duo.impl.DuoPrincipalSerializer" />
                	<bean class="net.shibboleth.idp.authn.principal.impl.IdPAttributePrincipalSerializer" />
                	<bean class="net.shibboleth.idp.authn.principal.impl.PasswordPrincipalSerializer" p:dataSealer="#{'%{idp.sealer.storeResource:}'.trim().length() > 0 ? getObject('shibboleth.DataSealer') : null}" />
            	</list>
        	</property>
    	</bean>`
4. /opt/shibboleth-idp/conf/attribute-resolver-social.xml
example of get email attribute:
        <AttributeDefinition id="mail" xsi:type="ScriptedAttribute">
	<DisplayName xml:lang="en">E-mail</DisplayName>
        <DisplayDescription xml:lang="en">E-Mail</DisplayDescription>
        <AttributeEncoder xsi:type="SAML1String" name="urn:mace:dir:attribute-def:mail" />
        <AttributeEncoder xsi:type="SAML2String" name="urn:oid:0.9.2342.19200300.100.1.3" friendlyName="mail" />
      <Script>
      <![CDATA[
       load("nashorn:mozilla_compat.js");
	importPackage(Packages.net.shibboleth.idp.attribute);
	importPackage(Packages.net.shibboleth.idp.authn);
        importPackage(Packages.org.slf4j);
        importPackage(Packages.java.util);
        importPackage(Packages.java.lang);
        importPackage(Packages.fi.okm.mpass.idp.authn.context.SocialUserContext);

	  authnContext = resolutionContext.getParent().getSubcontext("net.shibboleth.idp.authn.context.AuthenticationContext");
          socialContext = authnContext.getSubcontext("fi.okm.mpass.idp.authn.context.SocialUserContext");
          mail.addValue(socialContext.getEmail());
	
	  logger = LoggerFactory.getLogger("net.shibboleth.idp.attribute");
          logger.info("TESTE EMAIL" + socialContext.getEmail() );
          ]]>
        </Script>
    </AttributeDefinition>

Any help about I need to implement?

Thanks in advance.
Mauro

**Comment by hjmikkon** [comment]
Hi Mauro,

Apologies for a long delay, but I had extended summer leave.

When you're accessing the other service, the IdP's SSO session is used, which contains more limited information than the session where you authenticated, attributes were resolved, etc. With the SSO, you cannot find socialContext anywhere, you have to parse the attributes in the following way:

```      <ad:Script><![CDATA[
          authnContext = resolutionContext.getParent().getSubcontext("net.shibboleth.idp.authn.context.AuthenticationContext");
          subject = authnContext.getAuthenticationResult().getSubject();
          principals = subject.getPrincipals(Java.type("fi.okm.mpass.idp.authn.principal.SocialUserPrincipal").class);
          iterator = principals.iterator();
          while (iterator.hasNext()) {
              principal = iterator.next();
              if ("providerId".equals(principal.getType())) {
                  providerId.addValue(principal.getValue());
              }
          }
	]]></ad:Script>`

This assumes that you have configured principal serializers like you said in point 3. In this example, providerId is the attribute that is looked after, but you can change it to your desired claim.

**Comment by mjpmachado** [comment]
Hi Henri,

Thanks for the reply and sorry for my long delay too.
To resolve the problem we need only to configure the IdP property idp.session.enabled to false:

# Set to false to disable the IdP session layer
idp.session.enabled = false

Regards,
Mauro
